chia/full_node/full_node_api.py

```diff
@@ -1720,5 +1712,201 @@ def get_fee_estimates(est: FeeEstimatorInterface, req_times: List[uint64]) -> Li
         msg = make_msg(ProtocolMessageTypes.respond_fee_estimates, response)
         return msg
 
+    @api_request(
+        peer_required=True,
+        reply_types=[ProtocolMessageTypes.respond_add_puzzle_subscriptions],
+    )
+    async def request_add_puzzle_subscriptions(
+        self, request: wallet_protocol.RequestAddPuzzleSubscriptions, peer: WSChiaConnection
+    ) -> Message:
+        peer_id = peer.peer_node_id
+        max_subscriptions = self.max_subscriptions(peer)
+        subs = self.full_node.subscriptions
+
+        peak_height = self.full_node.blockchain.get_peak_height()
+        assert peak_height is not None
+
+        header_hash = self.full_node.blockchain.height_to_hash(peak_height)
+        assert header_hash is not None
+
+        new_subscriptions = set(request.puzzle_hashes) - subs.puzzle_subscriptions(peer_id)
+        added = subs.add_puzzle_subscriptions(peer_id, list(new_subscriptions), max_subscriptions)
+
+        response = wallet_protocol.RespondAddPuzzleSubscriptions(list(added), peak_height, header_hash)
+        msg = make_msg(ProtocolMessageTypes.respond_add_puzzle_subscriptions, response)
+        return msg
+
+    @api_request(
+        peer_required=True,
+        reply_types=[ProtocolMessageTypes.respond_add_coin_subscriptions],
+    )
+    async def request_add_coin_subscriptions(
+        self, request: wallet_protocol.RequestAddCoinSubscriptions, peer: WSChiaConnection
+    ) -> Message:
```

Comment:
here's a crazy idea; maybe there should be an option to automatically unsubscribe to coin IDs after they've been spent and the spend has been buried under some specified number of confirmations

Reply:
I don't like the idea of keeping track of an additional bit of data associated with each subscription. We do have the `RequestRemoveCoinSubscriptions` message now, so the wallet could unsubscribe from things as needed of its own volition. Do you think that's too roundabout a way to get there, though?

Reply:
the only argument in favour of auto-unsubscribe is to let the node clear out subscriptions by itself (and thus saving memory). it's not a strong case.

Reply:
I'm not sure it's worth the effort it would take to implement this, and every wallet could have different requirements for how it needs to handle subscriptions. I think using either an in memory list or a database to track subscriptions for each peer on the wallet side could work, or worst case you can now reset the subscriptions with the new protocol.